Code/Framework/AzFramework/AzFramework/Terrain/TerrainDataRequestBus.h

```diff
@@ -131,6 +133,41 @@ namespace AzFramework
                 Sampler sampleFilter = Sampler::DEFAULT,
                 bool* terrainExistsPtr = nullptr) const = 0;
 
+            //! Given a list of XY coordinates, call the provided callback function with surface data corresponding to each
+            //! XY coordinate in the list.
+            virtual void ProcessHeightsFromList(const AZStd::vector<AZ::Vector3>& inPositions,
+                SurfacePointListFillCallback perPositionCallback,
+                Sampler sampleFilter = Sampler::DEFAULT) const = 0;
+            virtual void ProcessNormalsFromList(const AZStd::vector<AZ::Vector3>& inPositions,
+                SurfacePointListFillCallback perPositionCallback,
+                Sampler sampleFilter = Sampler::DEFAULT) const = 0;
+            virtual void ProcessSurfaceWeightsFromList(const AZStd::vector<AZ::Vector3>& inPositions,
+                SurfacePointListFillCallback perPositionCallback,
+                Sampler sampleFilter = Sampler::DEFAULT) const = 0;
+            virtual void ProcessSurfacePointsFromList(const AZStd::vector<AZ::Vector3>& inPositions,
+                SurfacePointListFillCallback perPositionCallback,
+                Sampler sampleFilter = Sampler::DEFAULT) const = 0;
```

Comment:
I'm not as familiar with all the use cases for this as you are. So I'll leave it up to your judgement. I'll update this next week.

Reply:
I updated this to array_view and realized that array_view was in the AtomCore lib and not AzCore. This created a dependency on AtomCore(i.e., AzFramework now depends on AtomCore). Since a lot of other libraries and executables depend on AzFramework, I made some changes to the AtomCore libs to avoid linking AtomCore to a lot of other libs unnecessarily.

Reply:
Yeah, I've submitted a GHI feature request for someone to implement AZStd::span, which is the _real_ answer we want here, but for now it's an AtomCore dependency. :(

Reply:
If AZStd::array_view currently lives in AtomCore and we're jumping through hoops to make it available to other things, but we ultimately want to use AZStd::span instead, then it sounds like we're going to have an API change at some point anyway? If so I think we should just wait until AZStd::span gets implemented and use AZStd::vector (or whatever container makes sense) in the meantime, otherwise we're introducing complexity for something temporary without much to gain.

Reply:
I lean towards either using 
- array_view - It's the closest match for span which we want to move to, but requires a const cast under the hood.
- passing in start / end iterators directly - This gets us the functionality we need now without having to do anything nasty under the hood.

I don't think we want to force callers of this API to be using AZStd::vector<>.

Reply:
The const cast under the hood doesn't concern me at all (it's just an implementation detail), but having to add dependencies on AtomCore just to get the functionality of array_view is a sign to me that something is not right.

If we're not getting an AZStd::span anytime soon and really want to use array_view in the meantime, we should move it to AzCore so things that want to use it don't have to jump through CMake dependency hoops.

Reply:
The main thing I'm wanting to avoid is users of this interface having to architect around AZStd::vector when that might not be the right data structure for their system. I get the concern about AzFramework depending on AzCore, that's not good either. I don't think moving array_view is worth the effort though, that would require a lot of updates that will just need to be updated again later.

With all that in mind, I'd suggest either going with
- Pass in iterrators directly
- Create a AZStd::span class in AzCore (this could just be a copy of array_view with the names changed and allowing for non-const access, so shouldn't be hard)

Reply:
Maybe array_view isn't a part of AzCore because the C++ standard library doesn't have it either(MS does have its own implementation)? C++ currently only has span in the standard library. Not sure if moving it to AzCore might be a problem for that reason.

Reply:
I array_view is what span was originally going to be called and our implementation was based on an early spec for it, but then its name was changed to view to avoid confusion with something like string_view which does need to enforce const access.

Reply:
If we want to add span, we'd need to add support for `first()`, `last()`, and `subspan()` as well. Otherwise, it would be misleading.

Reply:
If the existing AZStd::array_view in AtomCore is already functionally equivalent to the desired AZStd::span in AzCore, I think creating the latter as a copy of the former would be fine.
- The implementation can always be refined later if needed.
- Things that want to use the functionality can start doing so using AZStd::span without having to depend on AtomCore to get at AZStd::array_view
- Things that are currently using AZStd::array_view from AtomCore don't need to be updated (although they can be over time and AZStd::array_view can be marked as deprecated in favour of AZStd::span).
- None of the APIs in this PR need to change later down the track (be that from AZStd::vector -> AZStd::span or AZStd::array_view -> AZStd::span).

Reply:
Agree with @bosnichd on this. I think it's ok to add support for `first()`, `last()`, and `subspan()` later as long as that work is tracked.

Reply:
Ok. I created a first pass of span copied from array_view and gave access to non-const iterators/pointers.

Reply:
My thoughts are to keep the `array_view.h` inside of AtomCore, but move it's implementation to `AzCore/std/containers/span.h`

Inside of `array_view.h` it would just be a using alias to `AZStd::span`

We should also add the constexpr variable of `inline constexpr std::size_t dynamic_extent = std::numeric_limits<std::size_t>::max();`

Reply:
I really like @lumberyard-employee-dm's idea if it works for everyone else?

Reply:
So I believe @mbalfour-amzn tried to just update array_view to allow for non-const access, but this had trickle down effects in Atom where the constness was expected. He initially attempted to just update places in Atom that required it, but it became such a large amount of work he pivoted away from that idea. So I think we'll either need to update all the places in Atom that expect const array_views, or just duplicate the implementation in the short term.

Reply:
Changing array_view to be `AZStd::span<const Element>` results in a bunch of errors in Atom along these lines:
`multiple overloads of 'validate_iterator' instantiate to the same signature`. This comes from vector.h where both validate_iterator(const iterator&) and validate_iterator(const const_iterator) evaluate to the same signature for array_view. There's also an issue with converting from an AZStd::array to array_view in SkinnedMeshInputBuffers.cpp:
```
no viable conversion from returned value of type 'const AZStd::array<RHI::Ptr<RHI::BufferView>,
      static_cast<uint8_t>(SkinnedMeshInputVertexStreams::NumVertexStreams)>' (aka 'const array<intrusive_ptr<AZ::RHI::BufferView>, static_cast<uint8_t>(SkinnedMeshInputVertexStreams::NumVertexStreams)>') to function return type
      'AZStd::array_view<AZ::RHI::Ptr<RHI::BufferView>>' (aka 'span<const AZStd::intrusive_ptr<AZ::RHI::BufferView>>')
```
I feel like getting array_view to be a const version of span is out of scope for this PR. I can add the partial version of span with non-const iterators/pointers to AzCore and update the ProcessList functions to use that. The proper implementations of `span` and `array_view` can be updated under the hood later without impacting any of the APIs that use them.

Reply:
The resaon depending on AtomCore wasn't bothering me personally too much is because it's not like we're depending on the Atom Gem, it's literally a folder that sits adjacent to AzCore in the core engine code at o3de/Code.  But I definitely agree that a _better_ approach is to make a first-pass version of AZStd::span and use that.  My main hesitation was that I thought creating it was going to be a largeish effort because I wasn't sure everyone would be OK with a partially-standards-compliant version that gets better later on.  I'd definitely rather see that than continue using array_view though.

Like @invertednormal said, I tried a couple days ago to make array_view work like span, and I changed 40+ files in Atom to add const into nearly every array_view, and it was still continuing to spiral outwards.  It's a big enough (and separate) task that I'd lean away from changing array_view at _all_ and just leave that as a separate cleanup task for the Atom folks.  I had high hopes for @lumberyard-employee-dm 's suggestion of adding the alias, since that's basically the change I was having to make everywhere, but there were definitely places where I was having to work through issues like the ones @amzn-sj lists, and it got tricky with things like array_view<AZ::RHI::Ptr<X>> for whether or not it becomes array_view<AZ::RHI::ConstPtr<const X>>, array_view<const AZ::RHI::Ptr<const X>>, etc.  I just wasn't sure what the intent was meant to be for some of the Atom APIs in terms of the exact const desirability.